ocaml/xapi/test_vdi_cbt.ml

```diff
@@ -234,79 +243,252 @@ let test_allowed_operations_updated_when_necessary () =
+      (__context, vDI, vbd)
+    in
 
-  OUnit.assert_equal ~msg:"VDI.data_destroy should destroy all associated VBDs"
-    (Db.VDI.get_VBDs ~__context ~self:vDI) [];
+    let test_data_destroy_succeeds =
+      (* The parameters tell us whether the VBD unplug operation has started/finished by the time we call VDI.data_destroy. *)
+      let test_data_destroy_succeeds ~vbd_unplug_started ~vbd_unplug_finished () =
+        let __context, vDI, vbd = setup_test () in
+        if vbd_unplug_started then start_vbd_unplug ~__context vbd;
+        if vbd_unplug_finished then finish_vbd_unplug ~__context vbd;
+
+        let _: Thread.t = Thread.create (fun () ->
+            if not vbd_unplug_started then begin
+              Thread.delay other_delay;
+              start_vbd_unplug ~__context vbd
+            end;
+            if not vbd_unplug_finished then begin
+              Thread.delay vbd_unplug_time;
+              finish_vbd_unplug ~__context vbd
+            end;
+            Thread.delay other_delay;
+            Db.VBD.destroy ~__context ~self:vbd
+          ) ()
+        in
+        let timeout = timeout +. timebox_timeout_delay in
+        Helpers.timebox
+          ~timeout
+          ~otherwise:(fun () -> failwith "data_destroy did not finish successfully in 5 seconds")
+          (fun () -> call_data_destroy ~__context ~self:vDI)
+      in
+      let open OUnit in
+      "test_data_destroy_succeeds" >:::
+      [ "test_data_destroy_succeeds_when_vbd_is_already_unplugged" >:: test_data_destroy_succeeds ~vbd_unplug_started:true ~vbd_unplug_finished:true
+      ; "test_data_destroy_succeeds_when_vbd_is_being_unplugged" >:: test_data_destroy_succeeds ~vbd_unplug_started:true ~vbd_unplug_finished:false
+      ; "test_data_destroy_succeeds_when_vbd_unplug_is_started_later" >:: test_data_destroy_succeeds ~vbd_unplug_started:false ~vbd_unplug_finished:false
+      ]
+    in
 
-  OUnit.assert_equal ~msg:"VM.suspend_VDI should be set to null"
-    (Db.VM.get_suspend_VDI ~__context ~self:vM) Ref.null;
+    let test_data_destroy_times_out =
+      (* The parameters tell us whether the VBD unplug operation has started/finished by the time VDI.data_destroy times out. *)
+      let test_data_destroy_times_out ~vbd_unplug_started ~vbd_unplug_finished () =
+        let __context, vDI, vbd = setup_test () in
+
+        let t = Thread.create (fun () ->
+            (* The data_destroy operation will time out in 4 seconds *)
+            let unplug_start_delay = match vbd_unplug_started, vbd_unplug_finished with
+              | true, true -> timeout -. vbd_unplug_time -. other_delay (* unplug starts & finishes before the timeout *)
+              | true, false -> timeout -. (vbd_unplug_time /. 2.0) (* unplug starts before & finishes after the timeout *)
+              | false, false -> timeout +. other_delay (* unplug starts after the 4s timeout *)
+              | _ -> failwith "not tested"
```

Comment:
Actually it's impossible that it has finished but not started, so I should say that here.

Reply:
Updated.